packages/execution-environments/src/common/endowments/network.ts

```diff
@@ -0,0 +1,343 @@
+import { allFunctions } from '../utils';
+
+type WebSocketCallback = (this: WebSocket, ev: any) => any;
+
+const createNetwork = () => {
+  // Open fetch calls or open body streams or open websockets
+  const openConnections = new Set<{ cancel: () => Promise<void> }>();
+
+  // Remove items from openConnections after they were garbage collected
+  const cleanup = new FinalizationRegistry<() => void>(
+    /* istanbul ignore next: can't test garbage collection without modifying node parameters */
+    (callback) => callback(),
+  );
+
+  const _fetch: typeof fetch = async (
+    input: RequestInfo,
+    init?: RequestInit,
+  ): Promise<Response> => {
+    const abortController = new AbortController();
+    if (init?.signal !== null && init?.signal !== undefined) {
+      const originalSignal = init.signal;
+      // Merge abort controllers
+      originalSignal.addEventListener(
+        'abort',
+        () => {
+          abortController.abort((originalSignal as any).reason);
+        },
+        { once: true },
+      );
+    }
+
+    let res: Response;
+    let openFethConnection: { cancel: () => Promise<void> } | undefined;
+    try {
+      const fetchPromise = fetch(input, {
+        ...init,
+        signal: abortController.signal,
+      });
+
+      openFethConnection = {
+        cancel: async () => {
+          abortController.abort();
+          try {
+            await fetchPromise;
```

Comment:
Yes, we don't know how the fetch is implemented. It might throw in the next event queue task instead of instantaneously. After returning from `teardown`, all code running inside Snaps is expected to have finished running. This await ensures that it really did finish.